src/peft/tuners/xlora/model.py

```diff
@@ -0,0 +1,345 @@
+from typing import List, Optional, Tuple, Union
+
+import torch
+import torch.nn as nn
+from transformers import PreTrainedModel
+
+from peft.tuners.lora.model import LoraModel
+from peft.utils.peft_types import PeftType
+
+from .. import lora
+from .classifier import InhibitorFlagPayload, Number, xLoRAClassifier
+from .config import xLoRAConfig
+from .insertion import BaseTunerWrapper, PeftModelWrapper, xLoRAConv2dLayer, xLoRAEmbeddingLayer, xLoRALinearLayer
+
+
+def convert_layers_to_xlora(
+    base: nn.Module,  # PeftModel
+    config: xLoRAConfig,
+) -> int:
+    """
+    Returns the number of swapped layers.
+    """
+    assert isinstance(base.base_model, lora.LoraModel)
+    total_swapped = 0
+
+    scaling_keys = None
+    for module in base.modules():
+        if isinstance(module, lora.LoraLayer):
+            if not scaling_keys:
+                scaling_keys = list(module.scaling.keys())  # NOTE(EricLBuehler): Python 3.7: dicts are ordered!
+
+        if isinstance(module, lora.Linear):
+            assert scaling_keys is not None
+            new_layer: Union[xLoRALinearLayer, xLoRAEmbeddingLayer, xLoRAConv2dLayer] = xLoRALinearLayer(
+                model=base,
+                target=module,
+                target_forward=module.forward,
+                layer_number=total_swapped,
+                config=config,
+            )
+            module.forward = new_layer.forward  # type: ignore[method-assign]
+            total_swapped += 1
+        elif isinstance(module, lora.Embedding):
+            assert scaling_keys is not None
+            new_layer = xLoRAEmbeddingLayer(
+                model=base,
+                target=module,
+                target_forward=module.forward,
+                layer_number=total_swapped,
+                config=config,
+            )
+            module.forward = new_layer.forward  # type: ignore[method-assign]
+            total_swapped += 1
+        elif isinstance(module, lora.Conv2d):
+            assert scaling_keys is not None
+            new_layer = xLoRAConv2dLayer(
+                model=base,
+                target=module,
+                target_forward=module.forward,
+                layer_number=total_swapped,
+                config=config,
+            )
+            module.forward = new_layer.forward  # type: ignore[method-assign]
+            total_swapped += 1
+
+    return total_swapped
+
+
+class xLoRAModel(LoraModel):
```

Comment:
For Python classes, I'd prefer to have them capitalized. Could you live with `XLoraModel` etc.?

Reply:
That sounds good! I have updated the code accordingly.